### PR TITLE
Add 'change_feed_mode' to 'query_items_change_feed' API

### PR DESCRIPTION
**Pull request by allenkim0129**
# Description

In this PR, we added support for `change_feed_mode` for `query_items_change_feed` API. For the `change_feed_mode` argument, users can use either string values(`LatestVersion` or `AllVersionsAndDeletes`) or Enum values(ChangeFeedMode.`LATEST_VERSION`or `ChangeFeedMode.ALL_VERSIONS_AND_DELETES`).  The Enum can be imported from `change_feed_state.py` like the following:
```
from azure.cosmos._change_feed.change_feed_state import ChangeFeedMode

...
create or set "created_container"
... 

# query change feed with 'ALL_VERSIONS_AND_DELETES' mode
query_iterable = created_container.query_items_change_feed(
    change_feed_mode=ChangeFeedMode.ALL_VERSIONS_AND_DELETES,
)
```

Currently the default value for the `change_feed_mode` is `LATEST_VERSION` mode. If the `change_feed_mode` argument were not passed, `query_items_change_feed` will use `LATEST_VERSION` by default. This mode was the only mode supported before we added the `change_feed_mdoe` argument. Now users can choose either option by passing `change_feed_mode`

#### ChangeFeedMode.LATEST_VERSION
With this mode, we return all items of the latest/current version from a container. Even if some items were updated after creation, it only returns the latest version of all created items from a specified point of time, which could be passed by the following arguments. (By default, the point in time will be `Now`):
- `is_start_from_beginning`: Query all items created from the beginning of container creation. (`is_start_from_beginning` and `start_time` are exclusive, and cannot be used together.)
- `start_time`: Query items created from the `start time`. The start time can be [datetime, 'Now', or 'Beginning']. (`is_start_from_beginning` and `start_time` are exclusive, and cannot be used together.)
- `continuation`: Query all items created from a `continuation` token, which was stored from previous query run.

example:
```
# query change feed from beginning
query_iterable = created_container.query_items_change_feed(
    is_start_from_beginning=True,
    change_feed_mode=ChangeFeedMode.LATEST_VERSION,
)

# query change feed from Now
query_iterable = created_container.query_items_change_feed(
    start_time="Now",
    change_feed_mode=ChangeFeedMode.LATEST_VERSION,
)

# query change feed from a continuation token
query_iterable = created_container.query_items_change_feed(
    continuation=continuation_token_1,
    change_feed_mode=ChangeFeedMode.LATEST_VERSION,
)

```

#### ChangeFeedMode.ALL_VERSIONS_AND_DELETES
With this option, we return all items of all versions including creates, updates, deletes. If an item was deleted , the deleted items will be included, and the metadata will specify that the item was deleted. Currently, we backend server only supports `continuation` token to specify the beginning time. If `is_start_from_beginning` or `start_time` were used, exceptions will be risen with error message. If `continuation` token was not passed, the point in time will be set to `Now`.
 
example:
```
# query change feed from a continuation token
query_iterable = created_container.query_items_change_feed(
    continuation=continuation_token_1,
    change_feed_mode=ChangeFeedMode.ALL_VERSIONS_AND_DELETES,
)
```

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py [hunk 1]
```diff
@@ -29,6 +29,9 @@
 from abc import ABC, abstractmethod
 from enum import Enum
 from typing import Optional, Union, List, Any, Dict, Deque
+import warnings
+
+from typing_extensions import Literal
```
Comment (allenkim0129) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py:34]: Unused import must be removed

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py [hunk 2]
```diff
@@ -96,7 +108,9 @@ def from_json(
                 raise ValueError("Invalid base64 encoded continuation string [Missing version]")
 
             if version == ChangeFeedStateVersion.V2.value:
-                return ChangeFeedStateV2.from_continuation(container_link, container_rid, continuation_json)
+                change_feed_mode = change_feed_state_context.get("changeFeedMode")
+                return ChangeFeedStateV2.from_continuation(
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py:112]: the continuation token will include changeFeedMode, so even if customer has passed in changeFeedMode, we probably should ignore it

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py [hunk 3]
```diff
@@ -208,6 +222,14 @@ def __init__(
         else:
             self._continuation = continuation
 
+        if change_feed_mode is None or change_feed_mode == ChangeFeedMode.LATEST_VERSION:
+            self._change_feed_mode = http_constants.HttpHeaders.IncrementalFeedHeaderValue
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py:226]: hmm, why changing the changeFeedMode to be the header value. Since this value will also be used in the serialization, so I think we should stick to the changeFeedMode (enum)

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py [hunk 4]
```diff
@@ -394,6 +411,7 @@ def from_initial_state(
                 raise ValueError("partitionKey is in the changeFeedStateContext, but missing partitionKeyFeedRange")
         else:
             # default to full range
+            warnings.warn("'feed_range' empty. Use full range by default.", UserWarning)
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py:414]: should it be info level?

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py [hunk 5]
```diff
@@ -349,7 +365,8 @@ def from_continuation(
             cls,
             container_link: str,
             container_rid: str,
-            continuation_json: Dict[str, Any]) -> 'ChangeFeedStateV2':
+            continuation_json: Dict[str, Any],
+            change_feed_mode: ChangeFeedMode) -> 'ChangeFeedStateV2':
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py:369]: the changeFeedMode should come from the continuation itself

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    """Validate keyword arguments(kwargs).
+    The values of keyword arguments must match the expect type and conditions. If the conditions do not match,
+    errors will be raised with the error messages and possible ways to correct the errors.
+
+    :param kwargs: Keyword arguments to verify for query_items_change_feed API
+    :keyword change_feed_mode: Must be one of the values in the Enum, 'ChangeFeedMode'.
+        If the value was 'ALL_VERSIONS_AND_DELETES', the following keywords cannot be used:
+            - 'partition_key_range_id'
+            - 'is_start_from_beginning'
+            - 'start_time'
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py:89]: if the changeFeedMode is allVersionsAndDeletes, start_time can still be used when the value is "NOW"? and also I the is_start_from_beginning is false?

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    :type change_feed_mode: ChangeFeedState
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py:90]: why change_feed_mode is type ChangeFeedState? should it be changeFeedMode?

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    :keyword partition_key_range_id: Deprecated Warning.
+    :type partition_key_range_id: str
+    :keyword is_start_from_beginning: Deprecated Warning. Cannot be used with 'start_time'.
+    :type is_start_from_beginning: bool
+    :keyword start_time: Must be in supported types.
+    :type start_time: Union[~datetime.datetime, Literal["Now", "Beginning"]]
+
+
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py:98]: nit: extra space?

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+            raise ValueError(
+                f"Invalid change_feed_mode was used: '{kwargs['change_feed_mode']}'."
+                f" Supported 'change_feed_modes' are {ChangeFeedMode.to_string()}.")
+
+        if change_feed_mode == ChangeFeedMode.ALL_VERSIONS_AND_DELETES:
+            if "partition_key_range_id" in kwargs:
+                raise ValueError(
+                    f"'{ChangeFeedMode.ALL_VERSIONS_AND_DELETES}' mode is not supported with 'partition_key_range_id'."
+                    f" Please use 'feed_range' instead.")
+            if "is_start_from_beginning" in kwargs:
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py:118]: Same comments as above:

if the changeFeedMode is allVersionsAndDeletes, start_time can still be used when the value is "Now"? and also I the is_start_from_beginning is false?

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py [hunk 10]
```diff
@@ -507,11 +508,13 @@ def query_items_change_feed(
             start_time: Optional[Union[datetime, Literal["Now", "Beginning"]]] = None,
             partition_key: PartitionKeyType,
             priority: Optional[Literal["High", "Low"]] = None,
+            change_feed_mode: Optional[ChangeFeedState] = None,
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py:511]: why changeFeedState as the type?

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py [hunk 11]
```diff
@@ -631,80 +665,55 @@ def query_items_change_feed( # pylint: disable=unused-argument
...
         if hasattr(response_hook, "clear"):
             response_hook.clear()
 
         if self.container_link in self.__get_client_container_caches():
             feed_options["containerRID"] = self.__get_client_container_caches()[self.container_link]["_rid"]
 
         result = self.client_connection.QueryItemsChangeFeed(
-            self.container_link, options=feed_options, response_hook=response_hook, **kwargs
-        )
+            self.container_link, options=feed_options, response_hook=response_hook)
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py:716]: why remove **kwargs

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py (hunk 4 above)
Comment (allenkim0129) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py]: Changed to info level

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py (hunk 10 above)
Comment (allenkim0129) [reply sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py]: Fixed to `ChangeFeedMode`

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py (hunk 11 above)
Comment (allenkim0129) [reply sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py]: Mb, I thought it was not used later in the code, but I found one place we used it. I put it back again

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py [hunk 12]
```diff
@@ -48,6 +49,18 @@ class ChangeFeedStateVersion(Enum):
...
 
+class ChangeFeedMode(str, Enum):
+    LATEST_VERSION = "LatestVersion"
+    ALL_VERSIONS_AND_DELETES = "AllVersionsAndDeletes"
+
+    def __str__(self):
+        return self.value
+
+    @classmethod
+    def to_string(cls):
```
Comment (annatisch) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py:60]: We don't need this.

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py [hunk 13]
```diff
@@ -631,71 +665,47 @@ def query_items_change_feed( # pylint: disable=unused-argument
...
+        :type change_feed_mode: Optional[ChangeFeedMode]
         :keyword Callable response_hook: A callable invoked with the response metadata.
         :param Any args: args
         :returns: An Iterable of items (dicts).
         :rtype: Iterable[Dict[str, Any]]
         """
         # pylint: disable=too-many-statements
-        if kwargs.get("priority") is not None:
-            kwargs['priority'] = kwargs['priority']
+        add_args_to_kwargs(args, kwargs)
```
Comment (annatisch) [inline sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py:683]: I don't believe this is needed here, as I don't think the async version ever supported `args` to begin with, so the backcompat shouldn't be necessary.

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+        1. partition_key_range_id
+        2. is_start_from_beginning
+        3. continuation
+        4. max_item_count
+    :type args: Tuple[Any, ...]
+    :param kwargs: Keyword arguments
+    :type kwargs: dict[str, Any]
+    """
+    if len(args) > 4:
+        raise ValueError(f"Too many arguments. Expected: less than 4, but given: {len(args)}")
```
Comment (annatisch) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py:58]: This should be:
`TypeError: foo() takes 4 positional arguments but {len(args)} were given`

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+            ('partition_key_range_id', str),
+            ('is_start_from_beginning', bool),
+            ('continuation', str),
+            ('max_item_count', int),
+        ]
+        for i, value in enumerate(args):
+            key, expected_type = key_and_types[i]
+
+            if key in kwargs:
+                raise ValueError(f"'{key}' is in both positional and keyword argument list. Please remove one of them.")
```
Comment (annatisch) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py:71]: This should be:
`TypeError: foo() got multiple values for argument '{key}'`

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+
+            if not isinstance(value, expected_type):
```
Comment (annatisch) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py:73]: Let's remove this. We don't do type checking in Python like this.

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+    # Filter items with value None
+    kwargs = {key: value for key, value in kwargs.items() if value is not None}
+
+    # Validate the keyword arguments
+    if "change_feed_mode" in kwargs:
+        change_feed_mode = kwargs["change_feed_mode"]
+        if change_feed_mode not in ChangeFeedMode:
+            raise ValueError(
+                f"Invalid change_feed_mode was used: '{kwargs['change_feed_mode']}'."
+                f" Supported 'change_feed_modes' are {ChangeFeedMode.to_string()}.")
```
Comment (annatisch) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py:108]: Once you've removed `to_string` - you replace this with: `", ".join([m.value for m in ChangeFeedMode])`

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+                f"'is_start_from_beginning' must be 'bool' type,"
+                f" but given '{type(kwargs['is_start_from_beginning']).__name__}'.")
+
+    elif "start_time" in kwargs:
+        start_time = kwargs['start_time']
+        if not isinstance(start_time, datetime) and not isinstance(start_time, str):
+            raise TypeError(
+                f"'start_time' must be either a 'datetime' or 'str' type, but given '{type(start_time).__name__}'.")
+
+        if isinstance(start_time, str) and start_time not in ["Now", "Beginning"]:
```
Comment (annatisch) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py:150]: I would make this check case-insensitive.
```suggestion
        if isinstance(start_time, str) and start_time.lower() not in ["now", "beginning"]:
```

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+            raise ValueError("'is_start_from_beginning' and 'start_time' are exclusive, please only set one of them.")
+
+        if not isinstance(kwargs["is_start_from_beginning"], bool):
+            raise TypeError(
+                f"'is_start_from_beginning' must be 'bool' type,"
+                f" but given '{type(kwargs['is_start_from_beginning']).__name__}'.")
+
+    elif "start_time" in kwargs:
+        start_time = kwargs['start_time']
+        if not isinstance(start_time, datetime) and not isinstance(start_time, str):
```
Comment (annatisch) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py:146]: Same comment here - we don't really do this kind of check in Python.

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+                raise ValueError(
+                    f"'{ChangeFeedMode.ALL_VERSIONS_AND_DELETES}' mode is not supported if 'partition_key_range_id'"
+                    f" was used. Please use 'feed_range' instead.")
+            if "is_start_from_beginning" in kwargs and kwargs["is_start_from_beginning"] is not False:
+                raise ValueError(
+                    f"'{ChangeFeedMode.ALL_VERSIONS_AND_DELETES}' mode is only supported if 'is_start_from_beginning'"
+                    f" is 'False'. Please use 'is_start_from_beginning=False' or 'continuation' instead.")
+            if "start_time" in kwargs and kwargs["start_time"] != "Now":
+                raise ValueError(
+                    f"'{ChangeFeedMode.ALL_VERSIONS_AND_DELETES}' mode is only supports if 'start_time' is 'Now'."
```
Comment (annatisch) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py:121]: I would rephrase to something more like:
`A 'start_time' value of '{start_time}' is incompatible with query mode '{mode}'.`

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py (hunk 15 above)
Comment (annatisch) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py:71]: This logic also doesn't work for detecting a conflict between `is_start_from_beginning` and `start_time`. (I would probably also do that conversion in this method to save you having to do the extra check in the `validate_kwargs` method.

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py [hunk 21]
```diff
@@ -521,9 +524,15 @@ def query_items_change_feed(
...
             (logical partition or a subset of a container)
         :type partition_key: Union[str, int, float, bool, List[Union[str, int, float, bool]]]
-        :keyword Literal["High", "Low"] priority: Priority based execution allows users to set a priority for each
+        :keyword priority: Priority based execution allows users to set a priority for each
             request. Once the user has reached their provisioned throughput, low priority requests are throttled
             before high priority requests start getting throttled. Feature must first be enabled at the account level.
+        :type priority: Optional[Literal["High", "Low"]]
+        :keyword change_feed_mode: The change feed mode enum to use when processing change feed items.
+            LATEST_VERSION: Query latest items from 'start_time' or 'continuation' token.
+            ALL_VERSIONS_AND_DELETES: Query all versions and deleted items from either 'fromNow'(default)
```
[the remark this answers is not in the document]
Comment (allenkim0129) [reply sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py:533]: I have verified by running Sphinx

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py [hunk 22]
```diff
@@ -507,11 +508,13 @@ def query_items_change_feed(
             start_time: Optional[Union[datetime, Literal["Now", "Beginning"]]] = None,
             partition_key: PartitionKeyType,
             priority: Optional[Literal["High", "Low"]] = None,
+            change_feed_mode: Optional[ChangeFeedMode] = None,
```
[the remark this answers is not in the document]
Comment (allenkim0129) [reply sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py]: Changed to `Optional[Literal["LatestVersion", "AllVersionsAndDeletes"]]`

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py (hunk 19 above)
Comment (allenkim0129) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py:144]: The type check for `datetime` was already there, so I didn't want to remove it.

The type check for `String` must exist to call `lower()` and check with compatible values. ("Now" & "Beginning"). 

Btw, I updated the order to made it more clear to understand

File: sdk/cosmos/azure-cosmos/azure/cosmos/container.py [hunk 23]
```diff
@@ -328,24 +329,33 @@ def query_items_change_feed(
...
-        :keyword Literal["High", "Low"] priority: Priority based execution allows users to set a priority for each
+        :paramtype partition_key: Union[str, int, float, bool, List[Union[str, int, float, bool]]]
+        :keyword priority: Priority based execution allows users to set a priority for each
             request. Once the user has reached their provisioned throughput, low priority requests are throttled
             before high priority requests start getting throttled. Feature must first be enabled at the account level.
-        :keyword Callable response_hook: A callable invoked with the response metadata.
+        :paramtype priority: Optional[Literal["High", "Low"]]
+        :keyword mode: The change feed mode enum to use when processing change feed items.
+            LATEST_VERSION: Query latest items from 'start_time' or 'continuation' token.
+            ALL_VERSIONS_AND_DELETES: Query all versions and deleted items from either 'fromNow'(default)
```
Comment (annatisch) [inline sdk/cosmos/azure-cosmos/azure/cosmos/container.py:354]:
```suggestion
            ALL_VERSIONS_AND_DELETES: Query all versions and deleted items from either a start_time of 'Now'` (default)
```

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+                raise ValueError(
+                    "'AllVersionsAndDeletes' mode is not supported if 'partition_key_range_id'"
+                    " was used. Please use 'feed_range' instead.")
+            if "is_start_from_beginning" in kwargs and kwargs["is_start_from_beginning"] is not False:
+                raise ValueError(
+                    "'AllVersionsAndDeletes' mode is only supported if 'is_start_from_beginning'"
+                    " is 'False'. Please use 'is_start_from_beginning=False' or 'continuation' instead.")
+            if "start_time" in kwargs and kwargs["start_time"] != "Now":
+                raise ValueError(
+                    "'AllVersionsAndDeletes' mode is only supports if 'start_time' is 'Now'."
```
Comment (annatisch) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py:119]: I would still make these checks case-insensitive.
```suggestion
                    "'AllVersionsAndDeletes' mode is only supported if 'start_time' is 'Now'."
```

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py (hunk 16 above)
Comment (allenkim0129) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py]: As we discussed offline, I will ignore type check here. We should assume users passes the right type.

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py (hunk 19 above)
Comment (allenkim0129) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py]: I will remove check for string type check, but I will leave datetime since datetime doesn't have `lower()`.

File: sdk/cosmos/azure-cosmos/azure/cosmos/container.py (hunk 23 above)
Comment (allenkim0129) [reply sdk/cosmos/azure-cosmos/azure/cosmos/container.py:354]: Done

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py (hunk 24 above)
Comment (allenkim0129) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_utils.py:119]: Done

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py [hunk 25]
```diff
@@ -623,78 +659,54 @@ def query_items_change_feed( # pylint: disable=unused-argument
...
             ~datetime.datetime: processing change feed from a point of time. Provided value will be converted to UTC.
             By default, it is start from current ("Now")
-        :type start_time: Union[~datetime.datetime, Literal["Now", "Beginning"]]
-        :keyword Literal["High", "Low"] priority: Priority based execution allows users to set a priority for each
+        :paramtype start_time: Union[~datetime.datetime, Literal["Now", "Beginning"]]
+        :keyword priority: Priority based execution allows users to set a priority for each
             request. Once the user has reached their provisioned throughput, low priority requests are throttled
             before high priority requests start getting throttled. Feature must first be enabled at the account level.
-        :keyword Callable response_hook: A callable invoked with the response metadata.
+        :paramtype priority: Optional[Literal["High", "Low"]]
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py:673]: I think mode should be mutually exclusive with continuation, correct? Because the ocntinuation determines the mode as well as start from already?

File: sdk/cosmos/azure-cosmos/azure/cosmos/container.py [hunk 26]
```diff
@@ -440,106 +471,68 @@ def query_items_change_feed(
...
 
         # pylint: disable=too-many-statements
-        if kwargs.get("priority") is not None:
-            kwargs['priority'] = kwargs['priority']
+        add_args_to_kwargs(args, kwargs)
+        validate_kwargs(kwargs)
         feed_options = build_options(kwargs)
 
         change_feed_state_context = {}
-        # Back compatibility with deprecation warnings for partition_key_range_id
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/azure/cosmos/container.py:470]: big fan of moving this logic elsewhere, good stuff

File: sdk/cosmos/azure-cosmos/samples/change_feed_management_async.py [hunk 27]
```diff
@@ -77,6 +77,51 @@ async def read_change_feed_with_start_time(container, start_time):
...
+        print(doc)
+
+    print('\nFinished reading all the change feed from continuation\n')
+
+async def delete_all_items(container):
+    print('\nDeleting all item\n')
+
+    async for item in container.query_items(query='SELECT * FROM c'):
+        # Deleting the current item
+        print(item)
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/samples/change_feed_management_async.py:96]: no need to print the item when deleting/ creating

File: sdk/cosmos/azure-cosmos/samples/change_feed_management.py [hunk 28]
```diff
@@ -73,6 +73,52 @@ def read_change_feed_with_start_time(container, start_time):
...
+        print(doc)
+
+    print('\nFinished reading all the change feed from continuation\n')
+
+def delete_all_items(container):
+    print('\nDeleting all item\n')
+
+    for item in container.query_items(query='SELECT * FROM c', enable_cross_partition_query=True):
+        # Deleting the current item
+        print(item)
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/samples/change_feed_management.py:92]: no need to print the item when deleting/ creating if we're about to see it in the CF

File: sdk/cosmos/azure-cosmos/test/test_change_feed.py [hunk 29]
```diff
@@ -218,14 +251,6 @@ def create_random_items(container, batch_size):
         # Should equal batch size
         assert totalCount == batchSize
 
-        # test an invalid value, Attribute error will be raised for passing non datetime object
-        invalid_time = "Invalid value"
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_change_feed.py:222]: why did we remove this validation for previously existing tests?

**Review by simorenoh — CHANGES_REQUESTED** [review]
Just left a couple comments on formatting stuff for the most part to keep the SDK aligned throughout - great job Allen! 👍

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py [hunk 30]
```diff
@@ -367,12 +365,18 @@ def from_continuation(
         if continuation_data is None:
             raise ValueError(f"Invalid continuation: [Missing {ChangeFeedStateV2.continuation_property_name}]")
         continuation = FeedRangeCompositeContinuation.from_json(continuation_data)
-        return ChangeFeedStateV2(
+
+        mode = continuation_json.get(ChangeFeedStateV2.mode_property_name)
+        if mode is None:
+            raise ValueError(f"Invalid continuation: [Missing {ChangeFeedStateV2.mode_property_name}]")
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py:371]: This would be breaking if someone presents a ChangeFeedStateV2 continuation from an earleir SDK version, correct? That is a breaking change - change feed is often long running - so deifnitely possible that continuation tokens form previous SDK versions are presented. If I am missing something and this would work somehow please add brief comment why it is safe and add a tests case where we capture the json of a continuation how it looked before this and then use that to validate old continuation tokens continue to work.

**Review by FabianMeiswinkel — CHANGES_REQUESTED** [review]
Looks good to me except for the comment in ChangeFeedStateV2

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py (hunk 30 above)
Comment (allenkim0129) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py:371]: It is safe as far as I know of. @xinlian12  Please correct me if I am wrong.

Basically, `from_continuation()` only gets called if the change feed state version was `V2` in continuation. Ever since `ChangeFeedStateV2` was added(by @xinlian12), it has been creating continuation tokens included the following 2 properties (https://github.com/Azure/azure-sdk-for-python/pull/36930/files#diff-2d46bcf89a9e926a031b2c470534f67c62d9da61175a8191b74c088ea8aa36aeR220)
- version_property_name = "v"
- mode_property_name = "mode"

If ChangeFeedState version was `v2`, continuation must contain `mode` property. If not, the continuation had been corrupted, we must raise an exception with missing property.

For older continuation tokens, the ChangeFeedState vesions shouldn't be v2, so they would never hit this point.

Also for the test cases, I already added the test with corrupted continuation token: https://github.com/Azure/azure-sdk-for-python/pull/38105/files#diff-8b3af0ea28342f506eb8698103b130f4c85beba28fbc17750506ba0359d30f48R435

File: sdk/cosmos/azure-cosmos/CHANGELOG.md [hunk 31]
```diff
@@ -22,12 +22,13 @@ This version and all future versions will support Python 3.13.
...
-* Added **provisional** helper APIs for managing session tokens. See [PR 36971](https://github.com/Azure/azure-sdk-for-python/pull/36971).
-* Added ability to get feed range for a partition key. See [PR 36971](https://github.com/Azure/azure-sdk-for-python/pull/36971).
+* Added Retry Policy for Container Recreate in the Python SDK. See [PR 36043](https://github.com/Azure/azure-sdk-for-python/pull/36043)
+* Added option to disable write payload on writes. See [PR 37365](https://github.com/Azure/azure-sdk-for-python/pull/37365)
+* Added get feed ranges API. See [PR 37687](https://github.com/Azure/azure-sdk-for-python/pull/37687)
+* Added feed range support in `query_items_change_feed`. See [PR 37687](https://github.com/Azure/azure-sdk-for-python/pull/37687)
+* Added **provisional** helper APIs for managing session tokens. See [PR 36971](https://github.com/Azure/azure-sdk-for-python/pull/36971)
+* Added ability to get feed range for a partition key. See [PR 36971](https://github.com/Azure/azure-sdk-for-python/pull/36971)
+* Added change feed mode support in `query_items_change_feed`. See [PR 38105](https://github.com/Azure/azure-sdk-for-python/pull/38105)
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/CHANGELOG.md:31]: please move this to the new changelog entry above, it doesn't belong with stuff we already released

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py (hunk 30 above)
Comment (allenkim0129) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py:373]: I also added the comment as you suggested

File: sdk/cosmos/azure-cosmos/CHANGELOG.md (hunk 31 above)
Comment (allenkim0129) [reply sdk/cosmos/azure-cosmos/CHANGELOG.md]: Good point! Done!

**Review by simorenoh — APPROVED** [review]

File: sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py (hunk 30 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py:373]: yea, if the customer is using the old continuationToken(like just the lsn), it will go through a different code path. 

when the new continuation token format being added, it always contain a "mode" property and default to LatestVersion. In this PR, the only thing changed is the value can be in the mode property: either LatestVersion or AllVersionsAndDeletes

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_change_feed/change_feed_state.py:373]: Ahh I missed line 221 in the old file - where mode was set to a constant. Makes sense.

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM-Thanks
